### ISSUE TITLE
Question: Return label map along with predictions

### ISSUE THREAD

**Issue by hampusrosvall** [triton-inference-server/server#3804]
I am migrating from TensorFlow Serving/TorchServe to Triton. My current deployment consists of a segmentation model that on the server side returns the predicted classes i.e., segment along with a label map to decipher the integer classes to string classes for further processing on client side. I would like to maintain this set up when migrating to Triton.

I have not yet found a way to accomplish this i.e., return the predictions along with a "custom object", that is, the label map on server side. Do you have any suggestions on how to move forward here?

Thanks

**Comment by dyastremsky** [comment]
Are you able to use the mapping as part of the config, like detailed [here?](https://github.com/triton-inference-server/server/blob/main/docs/model_management.md) Example config [here.](https://github.com/triton-inference-server/server/blob/main/docs/examples/model_repository/inception_graphdef/config.pbtxt#L17)

You'd then have the mapping, as your client could read the file.

**Comment by hampusrosvall** [comment]
@dyastremsky I have tried that throughout the day. The thing is that I have a convolutional segmentation model, and the response looks really weird. I experimented with a simple fully connected model that does binary classification, and that works fine. I then implemented a simple segmentation model that yields weird results.

Given the simple **fully connected model**
```python
input_ = Input(shape=(16,))
x = Dense(2)(input_)
x = Activation("softmax")(x)
model = Model(inputs=input_, outputs=x)
```

Using following configuration
```text
name: "model"
platform: "tensorflow_savedmodel"
max_batch_size: 4
input [
  {
    name: "input_1"
    data_type: TYPE_FP32
    dims: [ 16 ]
  }
]
output [
  {
    name: "activation"
    data_type: TYPE_FP32
    dims: [ 2 ]
    label_filename: "label_map.txt"
  }
]
default_model_filename: "model"
```

Yields results like this:
```text
output_array.shape (4, 2)
[b'0.500000:0:NEGATIVE' b'0.500000:1:POSITIVE']
```

Given a **convolutional segmentation model**
```python
input_ = Input(shape=(28, 28, 1))
x = Conv2D(32, kernel_size=(3, 3), padding="same")(input_)
x = Conv2D(2, kernel_size=(3, 3), padding="same")(x)
x = Activation("softmax")(x)
model = Model(inputs=input_, outputs=x)
```

Using following configuration
```text
name: "model"
platform: "tensorflow_savedmodel"
max_batch_size: 4
input [
  {
    name: "input_1"
    data_type: TYPE_FP32
    dims: [ 28, 28, 1 ]
  }
]
output [
  {
    name: "activation"
    data_type: TYPE_FP32
    dims: [ 28, 28, 2 ]
    label_filename: "label_map.txt"
  }
]
default_model_filename: "model"
```

Yields results like this
```text
output_array.shape (4, 2)
[b'0.500000:1041' b'0.500000:1053']
```

I use this gRPC client to perform the requests

```python
dense = True
inputs = []
outputs = []
shape = [4, 16] if dense else [4, 28, 28, 1]
inputs.append(grpcclient.InferInput("input_1", shape, "FP32"))
input0_data = np.zeros(shape=shape, dtype=np.float32)

# Initialize the data
inputs[0].set_data_from_numpy(input0_data)

outputs.append(grpcclient.InferRequestedOutput("activation", class_count=2))

# Test with outputs
results = triton_client.infer(
    model_name=model_name,
    inputs=inputs,
    outputs=outputs,
)

output_array = results.as_numpy("activation")
print(f"output_array.shape {output_array.shape}")
for i, results in enumerate(output_array):
    print(results)
    for result in results:
        if output_array.dtype.type == np.object_:
            cls = "".join(chr(x) for x in result).split(':')
        else:
            cls = result.split(':')
    print(cls)
    if i > 0: break
```

And this `label_map.txt`
```text
NEGATIVE
POSITIVE
```

**Comment by dyastremsky** [comment]
Is it possible you have more classes than those 2 labels could define? It looks to me like the outputs in the second example may be classes 1041 and 1053 respectively, not covered by the labels.

**Comment by rmccorm4** [comment]
I think it also may more more sense to have classification probabilities as a separate output and attach the label_filename to that output in config.pbtxt - I believe this is more standard practice.

Currently there is a single multi-dimensional output containing the label_filename which I'm not sure would work as expected.

Instead of a single [28, 28, 2] output where the third dim is the label - they could have one output of shape [28, 28] (no labels need to be associated with this output) and another output for the label of shape [2] with the label_filename associated with it - I think this may fix the issue. Though you would need to change your model to have multiple outputs to match the new config.

This issue may help: https://github.com/triton-inference-server/server/issues/1524

**Comment by hampusrosvall** [comment]
@rmccorm4 I think that is the case for image _classification_, however, I am doing image segmentation and am getting the exact behaviour as described in this issue here https://github.com/triton-inference-server/server/issues/3787. I.e., I am getting a single element array output when I request a class for each pixel `(i, j)`. You can replicate the error with the following model, config and label map.

```python
import tensorflow as tf
from tensorflow.keras.layers import Input, Conv2D, Activation, Dense
from tensorflow.keras import Model

# Segmentation model that outputs (N, 28 28, 2)
input_ = Input(shape=(28, 28, 1))
x = Conv2D(32, kernel_size=(3, 3), padding="same")(input_)
x = Conv2D(2, kernel_size=(3, 3), padding="same")(x)
x = Activation("softmax")(x)
model = Model(inputs=input_, outputs=x)


@tf.function(input_signature=[tf.TensorSpec([None, 28, 28, 1], tf.float32)])
def serving_default(x):
    scores = model.__call__(x)
    return {
        "scores": scores,
    }


serve = serving_default.get_concrete_function()
tf.saved_model.save(
    model,
    './label_map/model/1/model',
    signatures={
        "serving_default": serve
    }
)
```

```text
name: "model"
platform: "tensorflow_savedmodel"
max_batch_size: 1
input [
  {
    name: "x"
    data_type: TYPE_FP32
    dims: [ 28, 28, 1 ]
  }
]
output [
  {
    name: "scores"
    data_type: TYPE_FP32
    dims: [ 28, 28, 2 ]
    label_filename: "label_map.txt"
  }
]
default_model_filename: "model"
```

```text
NEGATIVE
POSITIVE
```

**Comment by dyastremsky** [comment]
Thank you kindly for the additional information. I've filed a report to investigate. What version of Triton are you running? Did you build it yourself or use the pre-built container?

**Comment by dyastremsky** [comment]
Please see https://github.com/triton-inference-server/server/issues/3787#issuecomment-1018742847. If that doesn't solve it, feel free to reopen this ticket.

**Comment by hampusrosvall** [comment]
@dyastremsky I am using this Docker image `nvcr.io/nvidia/tritonserver:21.12-py3`. In https://github.com/triton-inference-server/server/issues/3787 the developer has a single neuron output and hence it makes sense to put `class_count=1` in the output request. However, my toy model in https://github.com/triton-inference-server/server/issues/3804#issuecomment-1013865548 has two outputs and hence I should put `class_count=2`.

**Comment by dyastremsky** [comment]
Thanks for providing additional context! Let me tag @deadeyegoodwin, as he was able to propose the original fix and might have an additional recommendation.

**Comment by deadeyegoodwin** [comment]
class_count has nothing to do with the number of outputs. For each output you request you can decide to have that output returned as a raw tensor (the default) or you can ask for it to be returned as a ordered list of classifications. You use class_count parameter for this and the value of class_count indicates how many classifications to return for that output. The algorithm used is explained in https://github.com/triton-inference-server/server/blob/main/docs/protocol/extension_classification.md. If the output tensor is not one dimensional, then it is flattened in row-major order for the purposes of the classification algorithm. 

It is possible that having the option to choose a different classification algorithm for an output could be useful in your case and it would help if you would specify exactly what that algorithm should be.

**Comment by dyastremsky** [comment]
Due to inactivity, I'll close this ticket. The tickets that mention this one should have related feature requests.

Please re-open the issue if you would like to follow up with this issue.
